blockprotocol/blockprotocol
# Abstract Temporal Versioning away from packages

## 🌟 What is the purpose of this PR?

In previous PRs we introduced support for bi-temporal versioning (see #924, #928, #931). This is a powerful extension of the Graph module within the BP, but adds a large amount of complexity that can be avoided in the vast majority of cases. As such, this PR tries to reduce incidental exposure of users to this complexity if they do not need it. As such, most blocks should be able to be written without any knowledge that bitemporal versioning exists, and certain EAs need not support/implement it if they have no use for it (although doing so will limit the blocks they can support).

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1203358502199087/1203968498369973/f) _(internal)_
- [Asana task](https://app.asana.com/0/1203623179643284/1203968498369970/f) _(internal)_

## 🚫 Blocked by

N/A

## 🔍 What does this change?
(Please review commit by commit and take note of the messages, as various refactorings will make the diff much bigger otherwise)

- Restructures the inside of the Graph module to support more entrypoints
- Creates `temporal` and `non-temporal` variants for most of the entrypoints to the Graph module (except for `internal`, and `graph-module-json`)
- Reworks MBD to support temporal _and_ non-temporal versioning.
- Updates the dev app within MBD to be non-temporal by default

## 📜 Does this require a change to the docs?

- As with previous PRs, the 0.3 branch requires a large amount of changes to the docs. This is being done imminently, but not as part of this PR.

## ⚠️ Known issues

- If we try to keep `preserveModules` in the rollup config (to maintain the folder structure and not bundle all of the JS into one file) then there seems to be problems where various files / types / variables are incorrectly missed in bundling. I do not currently have capacity to look into this so we are producing bundled rolled up files. This is potentially/probably caused by `export *` not correctly rolling the elements in.
- We don't have a programmatic way of ensuring consistency of the two flavours of entrypoints
- Exporting an alias of a generic type, e.g. `type Foo = Bar<true>` breaks inference when that's used in another function,  
  - e.g. for `const useBar = <Baz extends boolean>(bar: Bar<Baz>): Baz` then:
    - `useBar(bar as Bar<true>): true`
    - `useBar(bar as Foo): boolean` (broken)
  - This caused a large number of problems when trying to use the `Subgraph` type, breaking functions like `getRoots`. This required redefining the types manually (as opposed to creating aliases) in the files that re-exported types. It's unclear if there are more problematic types, the `Entity` related ones with `Properties` might suffer the same fate.
- MBD heavily uses hooks. You cannot (should not) conditionally call a hook when rendering in React,  and as such a large quantity of the components and hooks had to be split (mostly duplicated) into two variants, `Temporal` and `NonTemporal`. We can probably deduplicate a lot of the code further with more work, but within the capacity I had at the time I went for the more brute-forcey approach of copy pasting.
- The `schema` step of the templates is currently broken due to a missing AJV dependency
    

## 🐾 Next steps

- Look into writing a lint to ensure the API surfaces are consistent/compatible

## 🛡 What tests cover this?

- `eslint` and `tsc` checks
- Manual checks of MBD

## ❓ How to test this?

1.  Checkout the branch / view the deployment
1.  Try running MBD as usual, ensure things work as expected
1.  To check temporal versioning works, checkout [Change dev-app.tsx to temporal versioning](https://github.com/blockprotocol/blockprotocol/pull/990/commits/ccdc2f16bd01d88233c2c2d014a1891b32e3f656) which applies the following:
    1.  Revert the changes to `dev-app` which removed temporal versioning information from the block entities. 
    1. Fix the imports to import from the "/temporal" entrypoint:
    ```
    import {
      EntityTemporalVersioningMetadata,
      QueryTemporalAxes,
    } from "@blockprotocol/graph/temporal";
    ```
    1.  Re-add `initialTemporalAxes: temporalAxes` to `initialData`
    1.  Change the `temporal={false}` prop to `temporal` in the `MockBlockDock` component
    1.  Check that MBD works as usual (with temporal versioning)

## 📹 Demo

### Non-Temporal
<img width="1712" alt="image" src="https://user-images.githubusercontent.com/25749103/220098832-a915e52d-6e43-48d4-914a-9f939a5bcae9.png">
<img width="834" alt="image" src="https://user-images.githubusercontent.com/25749103/220100043-dfa3896a-b4b0-48d2-bde7-16b4ed23d57a.png">

Note that there are still only 4 vertices within the datastore, despite the property being changed

### Temporal

<img width="1718" alt="image" src="https://user-images.githubusercontent.com/25749103/220099731-6e7e6e99-2a90-4b8d-8d38-8195758b9a15.png">
<img width="778" alt="image" src="https://user-images.githubusercontent.com/25749103/220099847-9e9348be-ab20-4e55-a722-cb5165867b80.png">

Note all the revisions in the datastore due to property changes
